### ISSUE TITLE
Improve parent finding algorithm with clustering

### ISSUE THREAD

**Issue by georgerichardson** [nestauk/rhodonite#23]
Current algorithm:

1. For communities at time period T, find most similar communities at T-1.
2. Find unions of the possible parent communities up to size n.
3. Find most similar community or union of communities, and label as parents.

With clustering:

1. For each time period, find all combinations of communities up to size n.
2. Cluster the communities and their combinations (hierarchically or OmniSlash)?
3. For a community at time period T, find a pool of potential parents in its community and those nearby.
4. Find most similar community or union of communities, and label as parents.

Potential advantages:

- Speed. Currently combinations of potential parent communities are calculated for each child community. This would calculate all combinations once in advance.
- Accuracy. The current algorithm only allows communities that are above a similarity threshold to be considered as candidates in the parent pool. This means that we are excluding the possibility that there are potential parents which are dissimilar to the child, until combined with other potential parents.

Potential disadvantages:

- Memory. Will need to compute combinations and cluster on large matrix of potential parents.
- Speed. Single calculation of combinations for all communities and clustering them may take a long time.

**Comment by georgerichardson** [comment]
Turns out this might be unfeasible as the number of combinations for each year could number in the millions, making the clustering slower than the overall running time of the algorithm.

A potential better option:

```
for clique_future in cliques_future:
    for element in clique_future:
        retrieve cliques from cliques_past that contains element
        find clique(s) that has closest similarity to clique_future
        append to list
    generate combinations of cliques
    find most likely parent
```

This should improve accuracy while drastically reducing the problem space for calculating the Jaccard similarity.

**Comment by georgerichardson** [comment]
closed with #26
